File: client/src/main/java/dev/throwouterror/game/renderer/MeshRenderer.java
/*
 * Copyright (c) Creepinson
 */

package dev.throwouterror.game.renderer;

import dev.throwouterror.game.object.Transform;
import org.lwjgl.opengl.GL11;
import org.lwjgl.opengl.GL30;

import static org.lwjgl.opengl.GL11.*;
import static org.lwjgl.opengl.GL15.*;

/**
 * @author Theo Paris
 */
public class MeshRenderer {
    private Transform transform;

    public MeshRenderer() {
        this.transform = new Transform();
    }

    public MeshRenderer(Transform transform) {
        this.transform = transform;
    }

    public void renderMesh(Mesh mesh) {
        GL11.glPushMatrix();
        GL11.glTranslated(this.transform.getPosition().x(), this.transform.getPosition().y(), this.transform.getPosition().z());
        GL11.glRotated(this.transform.getRotation().x(), 1.0, 0.0, 0.0);
        GL11.glRotated(this.transform.getRotation().y(), 0.0, 1.0, 0.0);
        GL11.glRotated(this.transform.getRotation().z(), 0.0, 0.0, 1.0);
        GL11.glScaled(this.transform.getScale().x(), this.transform.getScale().y(), this.transform.getScale().z());

        GL30.glBindVertexArray(mesh.getVAO());
        glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, mesh.getIBO());
        glVertexPointer(mesh.getVertices().length, GL_FLOAT, 0, 0L);
        glDrawElements(GL_TRIANGLES, mesh.getIndices().length, GL_FLOAT, 0);
        glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, 0);
        GL30.glBindVertexArray(0);

        GL11.glPopMatrix();
    }

    public Transform getTransform() {
        return transform;
    }
}

File: client/src/main/kotlin/dev/throwouterror/game/client/engine/renderer/MeshRenderer.kt
package dev.throwouterror.game.client.engine.renderer

import dev.throwouterror.game.client.engine.mesh.Mesh
import org.lwjgl.opengl.GL11
import org.lwjgl.opengl.GL15
import org.lwjgl.opengl.GL30

fun renderMesh(mesh: Mesh, vararg shaders: Shader) {
    GL30.glBindVertexArray(mesh.vao);
    GL30.glEnableVertexAttribArray(0);
    GL15.glBindBuffer(GL15.GL_ELEMENT_ARRAY_BUFFER, mesh.ibo);
    shaders.forEach { it.bind() }
    GL11.glDrawElements(GL11.GL_TRIANGLES, mesh.indicesCount, GL11.GL_FLOAT, 0);
    shaders.forEach { it.unbind() }
    GL15.glBindBuffer(GL15.GL_ELEMENT_ARRAY_BUFFER, 0);
    GL30.glDisableVertexAttribArray(0);
    GL30.glBindVertexArray(0);
}